File: tipgeometry.jl
abstract type AbstractTipGeometry end

struct Spherical{T<:AbstractFloat} <: AbstractTipGeometry
    R::T
end

struct Conical{T<:AbstractFloat} <: AbstractTipGeometry
    θ::T
end

function α(tip::Spherical{T})::T where {T<:AbstractFloat}
    return T(16 / 9) * sqrt(tip.R)
end

function α(tip::Conical{T})::T where {T<:AbstractFloat}
    return T(8.0 / (3.0π)) * tan(tip.θ)
end

function β(tip::Spherical{T})::T where {T<:AbstractFloat}
    return T(3 / 2)
end

function β(tip::Conical{T})::T where {T<:AbstractFloat}
    return T(2.0)
end

File: neuralconstitutive/tipgeometry.py
import abc

import equinox as eqx
import jax.numpy as jnp


class AbstractTipGeometry(eqx.Module):
    @abc.abstractmethod
    def a(self) -> float:
        pass

    @abc.abstractmethod
    def b(self) -> float:
        pass


class Spherical(AbstractTipGeometry):
    R: float

    def a(self) -> float:
        return (16 / 9) * jnp.sqrt(self.R)

    def b(self) -> float:
        return 1.5


class Conical(AbstractTipGeometry):
    theta: float

    def a(self) -> float:
        return (8 / (3 * jnp.pi)) * jnp.tan(self.theta)

    def b(self) -> float:
        return 2.0
